circuits/src/zk_circuits/valid_commitments.rs

```diff
@@ -0,0 +1,951 @@
+// | Circuit Definition |
+// ----------------------
+
+/// The circuit implementation of VALID COMMITMENTS
+pub struct ValidCommitments<
+    const MAX_BALANCES: usize,
+    const MAX_ORDERS: usize,
+    const MAX_FEES: usize,
+>;
+impl<const MAX_BALANCES: usize, const MAX_ORDERS: usize, const MAX_FEES: usize>
+    ValidCommitments<MAX_BALANCES, MAX_ORDERS, MAX_FEES>
+where
+    [(); MAX_BALANCES + MAX_ORDERS + MAX_FEES]: Sized,
+{
+    /// The circuit constraints for VALID COMMITMENTS
+    pub fn circuit<CS: RandomizableConstraintSystem>(
+        statement: ValidCommitmentsStatementVar,
+        mut witness: ValidCommitmentsWitnessVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES>,
+        cs: &mut CS,
+    ) {
+        // Reconstruct the base and augmented wallets
+        let recovered_blinder = witness.private_secret_shares.blinder.clone()
+            + witness.public_secret_shares.blinder.clone();
+        witness
+            .public_secret_shares
+            .unblind(recovered_blinder.clone());
+        witness.augmented_public_shares.unblind(recovered_blinder);
+
+        let base_wallet =
+            witness.private_secret_shares.clone() + witness.public_secret_shares.clone();
+        let augmented_wallet =
+            witness.private_secret_shares.clone() + witness.augmented_public_shares.clone();
+
+        // The mint that the wallet will receive if the order is matched
+        let mut receive_send_mint = CondSelectVectorGadget::select(
+            &[witness.order.quote_mint, witness.order.base_mint],
+            &[witness.order.base_mint, witness.order.quote_mint],
+            witness.order.side,
+            cs,
+        );
+        let receive_mint = receive_send_mint.remove(0);
+        let send_mint = receive_send_mint.remove(0);
+
+        // Verify that the wallets are the same other than a possibly augmented balance of
+        // zero for the received mint of the order. This augmented balance must come in place of
+        // a previous balance that was zero.
+        Self::verify_wallets_equal_with_augmentation(
+            statement.balance_receive_index,
+            receive_mint.clone(),
+            &base_wallet,
+            &augmented_wallet,
+            cs,
+        );
+
+        // Verify that the send balance is at the correct index
+        Self::contains_balance_at_index(
+            statement.balance_send_index,
+            witness.balance_send,
+            &augmented_wallet,
+            cs,
+        );
+        cs.constrain(witness.balance_send.mint - send_mint);
+
+        // Verify that the receive balance is at the correct index
+        Self::contains_balance_at_index(
+            statement.balance_receive_index,
+            witness.balance_receive,
+            &augmented_wallet,
+            cs,
+        );
+        cs.constrain(witness.balance_receive.mint - receive_mint);
+
+        // Verify that the fee balance is in the wallet
+        // TODO: Implement fees properly
```

Comment:
Fees will come in a follow up, still need to sketch out the details of how they fit into the design

Reply:
Naively, what's so different about fees? Can't we secret-split in the same way?

Reply:
Decrementing the fee is easy, you're correct. The problem is incrementing it elsewhere (i.e. the relayer and protocol wallets). It's not *hard* per se, it's just something I'm deferring until after the initial implementation to tame the complexity the first go around